rllib/utils/replay_buffers/multi_agent_episode_replay_buffer.py

```diff
@@ -0,0 +1,671 @@
+                        }
+                    )
+                # If sampled observation is terminal for the agent. Either MAE
+                # episode is truncated/terminated or SAE episode is truncated/
+                # terminated at this ts.
+                # TODO (simon, sven): Add method agent_alive(ts) to MAE.
+                # or add slicing to get_terminateds().
+                agent_ts = episode.env_t_to_agent_t[agent_id][episode_ts]
+                agent_eps = episode.agent_episodes[agent_id]
+                agent_terminated = agent_ts == agent_eps.t and agent_eps.is_terminated
+                agent_truncated = (
+                    agent_ts == agent_eps.t
+                    and agent_eps.is_truncated
+                    and not agent_eps.is_terminated
+                )
+                if episode_terminated or agent_terminated:
+                    is_terminated[module_id].append(True)
+                    is_truncated[module_id].append(False)
+                elif episode_truncated or agent_truncated:
+                    is_truncated[module_id].append(True)
+                    is_terminated[module_id].append(False)
+                else:
+                    is_terminated[module_id].append(False)
+                    is_truncated[module_id].append(False)
+
+            # Increase counter.
+            B += 1
+
+        # Should be convertible to MultiAgentBatch.
+        ret = {
+            **{
+                module_id: {
+                    Columns.OBS: batch(observations[module_id]),
+                    Columns.ACTIONS: batch(actions[module_id]),
+                    Columns.REWARDS: np.array(rewards[module_id]),
+                    Columns.NEXT_OBS: batch(next_observations[module_id]),
+                    Columns.TERMINATEDS: np.array(is_terminated[module_id]),
+                    Columns.TRUNCATEDS: np.array(is_truncated[module_id]),
+                    "weights": np.array(weights[module_id]),
+                    "n_steps": np.array(n_steps[module_id]),
+                }
+                for module_id in observations.keys()
+            }
+        }
+
+        # Return multi-agent dictionary.
+        return ret
+
+    def _num_remaining_episodes(self, new_eps, evicted_eps):
+        """Calculates the number of remaining episodes.
+
+        When adding episodes and evicting them in the `add()` method
+        this function calculates iteratively the number of remaining
+        episodes.
+
+        Args:
+            new_eps: List of new episode IDs.
+            evicted_eps: List of evicted episode IDs.
+
+        Returns:
+            Number of episodes remaining after evicting the episodes in
+            `evicted_eps` and adding the episode in `new_eps`.
+        """
+        return len(
+            set(self.episode_id_to_index.keys()).union(set(new_eps)) - set(evicted_eps)
+        )
+
+    def _evict_module_episodes(self, multi_agent_eps: MultiAgentEpisode) -> None:
```

Comment:
rename: `ma_episode` (let's use `sa_episode` and `ma_episode` for short var names).